typing/printtyp.ml

```diff
@@ -32,24 +32,255 @@ let rec longident ppf = function
+
+let map = Array.make Namespace.size M.empty
+let get namespace = map.(Namespace.id namespace)
+let set namespace x = map.(Namespace.id namespace) <- x
+
+(* Names used in recursive definitions are not considered when determining
+   if a name is already attributed in the current environment.
+   This is a weaker version of hidden_rec_items used by short-path. *)
+let protected = ref S.empty
+let add_protected id = protected := S.add (Ident.name id) !protected
+let reset_protected () = protected := S.empty
+
+let pervasives_name namespace name =
+  if not !enabled then Out_name.create name else
+  match M.find name (get namespace) with
+  | Associated_to_pervasives r -> r
+  | Need_unique_name _ -> Out_name.create (pervasives name)
+  | Uniquely_associated_to (id',r) ->
+      let hid, map = add_hid_id id' N.empty in
+      Out_name.set r (human_unique hid id');
+      Conflicts.explain namespace hid id';
+      set namespace @@ M.add name (Need_unique_name map) (get namespace);
+      Out_name.create (pervasives name)
+  | exception Not_found ->
+      let r = Out_name.create name in
+      set namespace @@ M.add name (Associated_to_pervasives r) (get namespace);
+      r
+
+(** Lookup for preexisting named item within the current {!printing_env} *)
+let env_ident namespace name =
+  if S.mem name !protected then None else
+  match Namespace.lookup namespace name with
+  | Pident id -> Some id
+  | _ -> None
+  | exception Not_found -> None
+
+(** Associate a name to the identifier [id] within [namespace] *)
+let ident_name_simple namespace id =
+  if not !enabled then Out_name.create (Ident.name id) else
+  let name = Ident.name id in
+  match M.find name (get namespace) with
+  | Uniquely_associated_to (id',r) when Ident.same id id' ->
+      r
+  | Need_unique_name map ->
+      let hid, m = find_hid id map in
+      Conflicts.explain namespace hid id;
+      set namespace @@ M.add name (Need_unique_name m) (get namespace);
+      Out_name.create (human_unique hid id)
+  | Uniquely_associated_to (id',r) ->
+      let hid', m = find_hid id' N.empty in
+      let hid, m = find_hid id m in
+      Out_name.set r (human_unique hid' id');
+      List.iter (fun (id,hid) -> Conflicts.explain namespace hid id)
+        [id, hid; id', hid' ];
+      set namespace @@ M.add name (Need_unique_name m) (get namespace);
+      Out_name.create (human_unique hid id)
+  | Associated_to_pervasives r ->
+      Out_name.set r ("Pervasives." ^ Out_name.print r);
+      let hid, m = find_hid id N.empty in
+      set namespace @@ M.add name (Need_unique_name m) (get namespace);
+      Out_name.create (human_unique hid id)
+  | exception Not_found ->
+      let r = Out_name.create name in
+      set namespace
+      @@ M.add name (Uniquely_associated_to (id,r) ) (get namespace);
+      r
+
+(** Same as {!ident_name_simple} but lookup to existing named identifiers
+    in the current {!printing_env} *)
+let ident_name namespace id =
```

Comment:
Why is this function needed? I'm refactoring the various lookup functions in the type-checker, and I'm not sure what to do with the one's used to implement `env_ident`.

The best I can work out is that this is trying to force us to print `t/2` when `t` has been shadowed even though we are not printing any references to the shadowing type. This doesn't seem desirable to me, especially printing the associated "conflict" explanation. It's also not implemented correctly anyway.

If it's not really needed I'll just remove it as part of my refactoring.

Reply:
This function is indeed trying to take in account names that have been already bound inside the environment. This is mostly useful when printing signature with `ocamlc -i` for cases like
```OCaml
type t = A
module M = struct
  type t
  let x = A
end
```
or having meaningful error message slightly earlier. For intance, the code below

```OCaml
type t = A
module M:sig type t val x:t end = struct type t let x = A end;;
```

raises

> Modules do not match:
         sig type t val x : t/2 end
       is not included in
         sig type t val x : t end
       Values do not match: val x : t/2 is not included in val x : t/1

without this function, the first part of the error message is useless 

> Modules do not match:
         sig type t val x : t end
       is not included in
         sig type t val x : t end
       Values do not match: val x : t/2 is not included in val x : t/1

If the feature is broken or too hard to implement, I imagine that it could be removed (temporarily or not).

Reply:
But why does this require checking for shadowing? I would have expected the naming context to contain all the `t`s that we print and so to add the extra markers in all those cases.

Besides:

```ocaml
Modules do not match:
sig type t val x : t/2 end
is not included in
sig type t val x : t end
Values do not match: val x : t/2 is not included in val x : t/1
```

doesn't seem to be much of an improvement: is `t/1` the same as one of the `type t`s or is it some third unrelated type?

Reply:
Because when printing a signature, we don't want to print extra-markers if they are not strictly required?

```OCaml
type t
module M = struct
  type t = A
  val x: t
end
```
should be printed as
```
type t
module M: sig
  type t = A
  val x: t
end
```
thus checking which `t` is in scope and which one is shadowed.